toolboxes/mri_image/hoMRImage.hxx

```diff
@@ -373,110 +373,109 @@ namespace Gadgetron
     }
 
     template <typename T, unsigned int D> 
-    bool hoMRImage<T, D>::serialize(char*& buf, size_t& len) const 
+    inline void hoMRImage<T, D>::printContent(std::ostream& os) const
     {
-        char* bufImage = NULL;
-        char* bufAttrib = NULL;
-
-        try
-        {
-            size_t lenImage(0);
-            GADGET_CHECK_THROW(BaseClass::serialize(bufImage, lenImage));
-
-            unsigned long long lenAttrib(0);
-
-            std::stringstream str;
-            ISMRMRD::serialize( const_cast<ISMRMRD::MetaContainer&>(attrib_), str);
-            std::string attribContent = str.str();
-            lenAttrib = attribContent.length()+1;
-
-            bufAttrib = new char[lenAttrib];
-            GADGET_CHECK_THROW(bufAttrib != NULL);
-
-            memset(bufAttrib, '\0', sizeof(char)*lenAttrib);
-            memcpy(bufAttrib, attribContent.c_str(), lenAttrib-1);
+        using namespace std;
+        BaseClass::printContent(os);
+        ISMRMRD::serialize( const_cast<ISMRMRD::MetaContainer&>(this->attrib_), os);
+    }
+}
 
-            size_t lenheader = sizeof(ISMRMRD::ISMRMRD_ImageHeader);
+template<class T, unsigned int D>
+void Gadgetron::Core::IO::read(std::istream &stream, Gadgetron::hoMRImage<T, D> &array)
```

Comment:
There is some redundancy between this and the read/write definitions for hoNDImage, is there a way to repurpose the hoNDImage image read/write inside these functions. It seems this code repetitions comes from the serialization format that you have chosen here. Since you own that contract, you could lay it out so that we reuse the hoNDImage read/write, I think? Would that break existing stuff?

Reply:
Actually the reason these functions are here is because we don't want primitives to depend on toolboxes. So primitives only have the basic types. For other types (e.g. hoMRImage), the read/write cannot be put into primitives without creating mutual reference. So I suggest we keep these functions in the hoMRImage.h/hxx.

Reply:
I meant that we can use the read/write for the primitive hoNDImage inside the read/write for hoMRImage since an hoMRImage is made up of an hoNDImage image and other "stuff" so there is now some repeated code in this functions, I am suggesting something where you use one inside the other and that may require you to change the way to serialize hoMRImage, but you own that contract.

Reply:
ok, use the primitive function calls to simplify the implementation.